nntrainer/src/bn_layer.cpp

```diff
@@ -32,28 +32,31 @@
 
 namespace nntrainer {
 
+enum class BNParams { mu, var, gamma, beta };
+
 /// @todo add channel wise bn for convolutional layer.
 int BatchNormalizationLayer::initialize(bool last) {
   int status = ML_ERROR_NONE;
 
-  dim = input_dim;
   dim = input_dim;
   dim.batch(1);
   output_dim = input_dim;
 
-  this->mu = Tensor(dim);
-  this->var = Tensor(dim);
-  this->gamma = Tensor(dim);
-  this->beta = Tensor(dim);
+  Tensor mu = Tensor(dim);
+  Tensor var = Tensor(dim);
+  Tensor gamma = Tensor(dim);
+  Tensor beta = Tensor(dim);
 
   mu.setZero();
   var.setValue(1);
   gamma.setZero();
   beta.setZero();
 
-  weights.clear();
-  weights.push_back(gamma);
-  weights.push_back(beta);
+  setParamSize(4);
+  paramsAt(0) = {std::move(mu), Tensor(dim), "BN:moving_average"};
```

Comment:
As not all params will use `apply_gradients` like in this case, how about later allowing not allocating memory for their grads?

Reply:
Yeah, do I 'have to' use pointer to make null kind of thing in c++?

Reply:
One way might be is to allow making empty tensors?

Reply:
Yup.

Reply:
I guess this can be done later.